mypy/checker.py

```diff
@@ -407,49 +407,63 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
+        if not defn.is_property:
+            self.check_overlapping_overloads(defn)
         return None
 
     def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
         # At this point we should have set the impl already, and all remaining
         # items are decorators
-        for i, item in enumerate(defn.items):
-            assert isinstance(item, Decorator)
-            sig1 = self.function_type(item.func)
-            for j, item2 in enumerate(defn.items[i + 1:]):
+        with experiments.strict_optional_set(True):
+            for i, item in enumerate(defn.items):
                 # TODO overloads involving decorators
-                assert isinstance(item2, Decorator)
-                sig2 = self.function_type(item2.func)
-                if is_unsafe_overlapping_signatures(sig1, sig2):
-                    self.msg.overloaded_signatures_overlap(i + 1, i + j + 2,
-                                                           item.func)
-            if defn.impl:
-                if isinstance(defn.impl, FuncDef):
-                    impl_type = defn.impl.type
-                elif isinstance(defn.impl, Decorator):
-                    impl_type = defn.impl.var.type
-                else:
-                    assert False, "Impl isn't the right type"
-                # This can happen if we've got an overload with a different
-                # decorator too -- we gave up on the types.
-                if impl_type is None or isinstance(impl_type, AnyType) or sig1 is None:
-                    return
+                assert isinstance(item, Decorator)
+                sig1 = self.function_type(item.func)
+
+                for j, item2 in enumerate(defn.items[i + 1:]):
+                    assert isinstance(item2, Decorator)
+                    sig2 = self.function_type(item2.func)
+
+                    assert isinstance(sig1, CallableType)
+                    assert isinstance(sig2, CallableType)
+
+                    if not are_argument_counts_overlapping(sig1, sig2):
+                        continue
 
-                assert isinstance(impl_type, CallableType)
-                assert isinstance(sig1, CallableType)
-                if not is_callable_subtype(impl_type, sig1, ignore_return=True):
-                    self.msg.overloaded_signatures_arg_specific(i + 1, defn.impl)
-                impl_type_subst = impl_type
-                if impl_type.variables:
-                    unified = unify_generic_callable(impl_type, sig1, ignore_return=False)
-                    if unified is None:
-                        self.fail("Type variable mismatch between " +
-                                  "overload signature {} and implementation".format(i + 1),
-                                  defn.impl)
+                    if overload_can_never_match(sig1, sig2):
+                        self.msg.overloaded_signature_will_never_match(
+                            i + 1, i + j + 2, item2.func)
```

Comment:
It's because we're iterating over the remaining overloads using an enumeration: 'j' starts at zero.

(`i + j + 2` was also what the code appeared to be doing before, so I figured it'd be fine to just copy it for the new check.)